src/scr.c

```diff
@@ -2358,7 +2376,7 @@ int SCR_Finalize()
   }
 
   /* flush checkpoint set if we need to */
-  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
+  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id) && !poststage_skip) {
```

Comment:
There's one additional case we want to add here.  It's possible that SCR hasn't started the flush of the most recent checkpoint.  For example, one might set ``SCR_FLUSH > 1`` and then call ``SCR_Finalize`` immediately after a writing a checkpoint that SCR didn't initiate a flush for.  SCR currently accounts for such a checkpoint by calling ``scr_flush_sync`` here.  We want to extend this so that if they are using poststage and if that checkpoint is in BBAPI, we call ``scr_flush_async_start`` instead of ``scr_flush_sync``.

Reply:
To expand on that, if ``SCR_FLUSH=2``, then on a new run, SCR would flush the following:
```
dataset id 1 = not flushed
dataset id 2 = async flush started in ``SCR_Complete_output``
dataset id 3 = not flushed
SCR_Finalize()
```
Given a pattern like that, in ``SCR_Finalize`` dataset 2 should be the ongoing async flush.  SCR would wait on dataset 2 to finish by calling ``scr_flush_async_wait`` and then call ``scr_flush_sync`` to flush dataset 3 (which happens to be the most recent checkpoint).

We'd want to initiate the flush of dataset 3 by calling ``scr_flush_async_start`` instead.

Reply:
Thanks for pointing this out.  I'm thinking something like?:
```C
  /* flush checkpoint set if we need to */
  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
    int flush_rc;

    if (scr_my_rank_world == 0) {
      scr_dbg(2, "Begin flushing checkpoints that haven't started @ %s:%d", __FILE__, __LINE__);
    }
    if (!poststage_skip) {
      flush_rc = scr_flush_sync(scr_cindex, scr_ckpt_dset_id);
    } else { 
      flush_rc = scr_flush_async_start(scr_ckpt_dset_id);
    }
  }
  if (flush_rc != SCR_SUCCESS) {
    scr_abort(-1, "Flush of dataset %d failed @ %s:%d",
        scr_ckpt_dset_id, __FILE__, __LINE__
        );
  }
```

Reply:
Yeah, that should work I think.  The double negative in ``!poststage_skip`` keeps twisting my mind, though.  Let's rename this variable to ``poststage`` and flip the meaning to be positive:
```
if (poststage) {
  scr_flush_async_start
} else {
  scr_flush_sync
}
```

Reply:
When using poststage, we'll also eventually want to change things to avoid waiting on dataset 2 to finalize in this example.

Reply:
Oh, I just remembered.  If also having dataset 2 finalize in the poststage script would require making more extensive changes in SCR to allow for multiple outstanding flushes at once, let's delay that extra bit to a future PR.  That may be a bigger change than we want to bite off in this PR.

Reply:
Would that mean the `SCR_Finalize()` behavior would change to a more general "we don't wait on any async, BBAPI-post-stageable transfers?" So with your example, it would mean that `SCR_Finalize()` doesn't wait on dataset 2, and would start dataset 3, and both 2 & 3 would get finalized in `scr_poststage.sh`?

Reply:
Ah ok, just saw your last comment after I had already posted.

Reply:
I'll keep things as we talked about earlier:
```C
  /* flush checkpoint set if we need to */
  if (scr_flush > 0 && scr_flush_file_need_flush(scr_ckpt_dset_id)) {
    int flush_rc;

    if (scr_my_rank_world == 0) {
      scr_dbg(2, "Begin flushing checkpoints that haven't started @ %s:%d", __FILE__, __LINE__);
    }

    if (poststage) {
      if (scr_flush_file_is_flushing(scr_ckpt_dset_id)) {
        flush_rc = SCR_SUCCESS;
      } else {
        flush_rc = scr_flush_async_start(scr_cindex, scr_ckpt_dset_id);
      }
    } else {
      /* Normal case */
      flush_rc = scr_flush_sync(scr_cindex, scr_ckpt_dset_id);
    }

    if (flush_rc != SCR_SUCCESS) {
      scr_abort(-1, "Flush of dataset %d failed @ %s:%d",
          scr_ckpt_dset_id, __FILE__, __LINE__
          );
    }
  }
```